datafusion/physical-expr/src/functions.rs

```diff
@@ -64,6 +65,22 @@ pub fn create_physical_expr(
 
     let data_type = function::return_type(fun, &input_expr_types)?;
 
+    // if function will be converted to case expression
+    if let BuiltinScalarFunction::If = fun {
+        if input_phy_exprs.len() != 3 {
+            return Err(DataFusionError::Internal(format!(
+                "{:?} args were supplied but IF function takes exactly 3 args",
+                input_phy_exprs.len(),
+            )));
+        }
+        let (when_expr, then_expr, else_expr) = (
+            Arc::clone(&input_phy_exprs[0]),
+            Arc::clone(&input_phy_exprs[1]),
+            Arc::clone(&input_phy_exprs[2]),
+        );
+        return case(None, vec![(when_expr, then_expr)], Some(else_expr));
```

Comment:
I would like to transfer the conversion code from its current location (where the physical expression was created) to a new analyzer rule. The new rule should happen before the `TypeCoercion` rule so that `expr[1]`, `expr[2]` can be correctly type coerced; by doing this, your test against Int32 and Int64 could be cast into common types. Cc @jackwener for expert opinion.

For example:
```rust
impl Analyzer {
    /// Create a new analyzer using the recommended list of rules
    pub fn new() -> Self {
        let rules: Vec<Arc<dyn AnalyzerRule + Send + Sync>> = vec![
            Arc::new(InlineTableScan::new()),
            Arc::new(RewriteIfExpr::new()), // rule that rewrite if expr
            Arc::new(TypeCoercion::new()),
            Arc::new(CountWildcardRule::new()),
        ];
        Self::with_rules(rules)
    }
```

Reply:
Good idea, let me check how to do it:)